subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/BeanNode.java

```diff
@@ -19,32 +19,132 @@
 import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Iterators;
+import org.gradle.api.Named;
 
 import javax.annotation.Nullable;
+import java.util.HashSet;
 import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
 
-public class BeanNode extends AbstractPropertyNode<BeanNode> {
-    private final Object bean;
+public abstract class BeanNode extends AbstractPropertyNode<BeanNode> {
 
-    public BeanNode(@Nullable String propertyName, Object bean) {
+    public static BeanNode create(@Nullable String propertyName, Object bean) {
+        if (bean instanceof Map<?, ?>) {
+            return new MapBeanNode(propertyName, (Map<?, ?>) bean);
+        }
+        if (bean instanceof Iterable<?>) {
+            return new IterableBeanNode(propertyName, (Iterable<?>) bean);
+        }
+        return new SimpleBeanNode(propertyName, bean);
+    }
+
+    protected BeanNode(@Nullable String propertyName, Object bean) {
         super(propertyName, Preconditions.checkNotNull(bean, "Null is not allowed as nested property '" + propertyName + "'").getClass());
-        this.bean = bean;
     }
 
-    public Object getBean() {
-        return bean;
+    public abstract Object getBean();
+
+    protected BeanNode createChildNode(String propertyName, @Nullable Object input) {
+        String qualifiedPropertyName = getQualifiedPropertyName(propertyName);
+        Object bean = Preconditions.checkNotNull(input, "Null is not allowed as nested property '" + qualifiedPropertyName + "'");
+        return BeanNode.create(qualifiedPropertyName, bean);
+    }
+
+    private static class SimpleBeanNode extends BeanNode {
+
+        private final Object bean;
+
+        public SimpleBeanNode(@Nullable String propertyName, Object bean) {
+            super(propertyName, bean);
+            this.bean = bean;
+        }
+
+        @Override
+        public Object getBean() {
+            return bean;
+        }
+
+        @Override
+        public boolean isIterable() {
+            return false;
+        }
+
+        @Override
+        public Iterator<BeanNode> getIterator() {
+            throw new UnsupportedOperationException();
+        }
```

Comment:
I'm a bit unhappy that we need two methods to handle this. Could we change these two methods to something like `addToQueue()`? We might be able to save the creation of an `Iterator` in places, too.

Reply:
One method it is.

Reply:
I like the one method, but I don't think it should be called `unpack`: we don't always have something to unpack, it's just what some of the implementations will do. Why not call it `addToQueue()`?

Reply:
Maybe `enqueue()`?